src/node_contextify.cc

```diff
@@ -1391,6 +1395,290 @@ void MicrotaskQueueWrap::RegisterExternalReferences(
+  args.GetIsolate()->CancelTerminateExecution();
+}
+
+void LocalWorker::Load(const FunctionCallbackInfo<Value>& args) {
+  LocalWorker* self = Unwrap(args);
+  if (self == nullptr) return;
+  if (!args[0]->IsFunction()) {
+    return THROW_ERR_INVALID_ARG_TYPE(
+        Environment::GetCurrent(args),
+        "The load() argument must be a function.");
+  }
+  Local<Value> result;
+  if (self->Load(args[0].As<Function>()).ToLocal(&result)) {
+    args.GetReturnValue().Set(result);
+  }
+}
+
+void LocalWorker::RunInCallbackScope(
+    const FunctionCallbackInfo<Value>& args) {
+  LocalWorker* self = Unwrap(args);
+  if (self == nullptr) return;
+  if (!args[0]->IsFunction()) {
+    return THROW_ERR_INVALID_ARG_TYPE(
+        Environment::GetCurrent(args),
+        "The runInCallbackScope() argument must be a function");
+  }
+  Local<Value> result;
+  if (self->RunInCallbackScope(args[0].As<Function>()).ToLocal(&result)) {
+    args.GetReturnValue().Set(result);
+  }
+}
+
+MaybeLocal<Value> LocalWorker::RunInCallbackScope(Local<Function> fn) {
+  if (context_.IsEmpty() || signaled_stop_) {
+    Environment* env = Environment::GetCurrent(isolate_);
+    THROW_ERR_INVALID_STATE(env, "Worker has been stopped");
+    return MaybeLocal<Value>();
+  }
+  LocalWorkerScope worker_scope(this);
+  v8::Isolate* isolate = isolate_;
+  CallbackScope callback_scope(isolate, wrap_.Get(isolate_), {1, 0});
+  MaybeLocal<Value> ret = fn->Call(context(), Null(isolate), 0, nullptr);
+  if (signaled_stop_) {
+    isolate->CancelTerminateExecution();
+  }
+  return worker_scope.EscapeMaybe(ret);
+}
+
+void LocalWorker::Start() {
+  signaled_stop_ = false;
+  Local<Context> outer_context = outer_context_.Get(isolate_);
+  Environment* outer_env = GetCurrentEnvironment(outer_context);
+  assert(outer_env != nullptr);
+  uv_loop_t* loop = GetCurrentEventLoop(isolate_);
+  assert(loop != nullptr);
+
+  MicrotaskQueue* microtask_queue =
+    outer_context_.Get(isolate_)->GetMicrotaskQueue();
+
+  Local<Context> context = Context::New(
+      isolate_,
+      nullptr /* extensions */,
+      MaybeLocal<ObjectTemplate>() /* global_template */,
+      MaybeLocal<Value>() /* global_value */,
+      DeserializeInternalFieldsCallback() /* internal_fields_deserializer */,
+      microtask_queue);
+  context->SetSecurityToken(outer_context->GetSecurityToken());
+  if (context.IsEmpty() || !InitializeContext(context).FromMaybe(false)) {
+    return;
+  }
+
+  context_.Reset(isolate_, context);
+  Context::Scope context_scope(context);
+  isolate_data_ = CreateIsolateData(
+      isolate_,
+      loop,
+      GetMultiIsolatePlatform(outer_env),
+      GetArrayBufferAllocator(GetEnvironmentIsolateData(outer_env)));
+  assert(isolate_data_ != nullptr);
+  ThreadId thread_id = AllocateEnvironmentThreadId();
+  auto inspector_parent_handle = GetInspectorParentHandle(
+      outer_env, thread_id, "file:///synchronous-worker.js");
+  env_ = CreateEnvironment(isolate_data_,
```

Comment:
This doesn't seem to be a good idea - I think so far we've been assuming in the code base that there is a 1:1 relationship between the environment and the isolate, things can be subtly broken once we have n:1. It would be cleaner if we follow the Realm approach and just split out states in the Environment that are unique to each context/realm to a subclass of `node::Realm` (or maybe it is already just `node::PrincipalRealm`).

Reply:
>  I think so far we've been assuming in the code base that there is a 1:1 relationship between the environment and the isolate, things can be subtly broken once we have n:1.

Quite a few of the crashes I've been fighting with this approach are due to that. I'm using the reference to `env` to clean up all the handles. Otherwise, we will have bad crashes after `stop()` is called (which is why I'm doing this in core). Using a `node::PrincipalReam` would require removing the deliberate `stop()` method: how would we shut it down?

The other question I have if we dith the `CreateEnvironment` call, how do we guarantee some level of isolation?

Reply:
You can just move the handles into the Realm instead. Basically, just move things that should belong to individual realms instead of being shared across them to the realm, and do the setup/cleanup on a per-realm basis, which is what we've been trying to do with the ShadowRealm integration.

Reply:
> You can just move the handles into the Realm instead.

How would you do that?

As an example `HandleWrap` add things to the `Environment` here:
https://github.com/nodejs/node/blob/c542d3a1d30d3e6c22b27f9bde55656923874818/src/handle_wrap.cc#L110 (there are a few other places too).

Should I try to move that list from the Env to the Realm?

Reply:
Yes, just make that `realm()->handle_wrap_queue()->PushBack(this)`. That's what we've been doing to support other BaseObjects in ShadowRealms, we just haven't got to `HandleWrap` yet.

Reply:
Moving the list from the Environment to the Realm might break the postmortem diagnostics since tools like llnode are built on top of the `Environment::handle_wrap_queue` structure: https://github.com/nodejs/node/blob/main/src/node_postmortem_metadata.cc#L23

An option can be tracking the handle wraps and request wraps by both the Env and its creation Realm.

Reply:
I think we don't need to worry much about the postmortem diagnostics data, we only need to leave the offset of the queues within Realms and the tools would then figure out how to adapt to newer versions of Node.js. It's never guaranteed that we'd never change the layout of our internals, only that when we do, we still leave some information in the binary (the offsets) for these tools to figure out how to extract information from a core dump / process memory.

Reply:
My understanding is that we won't be able to have a `process` object without a new Node.js Environment, right?

May we land this PR and work to remove `CreateEnvironment` in follow-on work?

Reply:
> That that's not the case is precisely the difference between Environment and IsolateData, though. Just because the Node.js CLI doesn't create multiple Environment instances per Isolate doesn't mean that it's not semantically correct to do so.

I don't think in practice we have really been writing the code that way. There are some places where we configure V8 isolate hooks (e.g. the ones in `Environment::InitializeDiagnostics`) with the current Environment as data to be used in the callback. The async hooks where things like `Environment::GetCurrent(isolate)->trigger_async_id()` is used a lot don't look particularly robust against a multi-Environment architecture either, and there are probably more places than I can think of off the top of my head..